[FLINK-15834] Set up nightly builds in Azure & various CI improvements

rmetzger opened the pull request:
## What is the purpose of the change

This change set up a nightly end to end test, testing special scenarios such as Java 11, hadoop 2.8 or scala 2.12.


## Brief change log

- Change the `build-apache-repo.yaml` to trigger the tests nightly (I left out release branches for now, as only master has the azure files present)
- change the `job-templates.yml` to build the end to end tests with every commit (pr, push builds). This is an experiment, as we might not have enough resources (but I believe we do).
I had two options for this: either transfer the build artifact from the compile phase into the e2e test job, or build there from scratch. I decided for the latter, as the build always takes 20 minutes. So in case the test machines are busy, but the azure provided machines are available, we'll start compiling right away.
The precommit tests are executed in this stage, before the end to end tests.

This change also fixes many issues with the end to end tests:
 
**Fix Kubernetes E2E tests**

Problem: Low disk space was causing K8s to mark the kubelet as "full disk", thus Flink did not schedule there.
Problem: Low disk space let the kublet delete unused docker images, including images generated for the test.

**Fix Kerberized YARN e2e test**

The problem was that YARN was decommissioning NodeManagers because of low disk space.

**Fix queryable state e2e test**

Problem: The logging pattern recently changed, that's why the extaction of port / ip failed


## Verifying this change

This nightly build failed only because of known issues (which also caused the travis nightly test to fail). I believe the structure / environment of the individual nightly tests is fine: https://dev.azure.com/rmetzger/Flink/_build/results?buildId=5590&view=results


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

rmetzger commented on flink-end-to-end-tests/test-scripts/common_yarn_docker.sh:
```diff
@@ -129,21 +129,31 @@ END
     docker exec master bash -c "kinit -kt /home/hadoop-user/hadoop-user.keytab hadoop-user"
     docker exec master bash -c "yarn application -list -appStates ALL"
     application_id=`docker exec master bash -c "yarn application -list -appStates ALL" | grep "Flink" | grep "cluster" | awk '{print \$1}'`
+    
```
I added these blank lines to have more visual structure in this method (which as grown a bit)

rmetzger commented on tools/azure-pipelines/prepare_precommit.sh:
```diff
@@ -32,8 +32,17 @@ find . -type f -name '*.timestamp' | xargs touch
+fi
 
 chmod -R +x build-target
 chmod -R +x flink-end-to-end-tests
```
I left this file in, even though it is currently not used. I consider always running the e2e tests an experiment. If we need to undo it, this file is still around.
If the experiment is successful, I will remove it (a cleanup PR is needed anyways once we remove travis support).

zentol commented on flink-end-to-end-tests/test-scripts/common_kubernetes.sh:
```diff
@@ -51,10 +51,18 @@ function check_kubernetes_status {
+                --extra-config=kubelet.minimum-container-ttl-duration=120m \
+                --extra-config=kubelet.eviction-hard="memory.available<5Mi,nodefs.available<1Mi,imagefs.available<1Mi" \
+                --extra-config=kubelet.eviction-soft="memory.available<5Mi,nodefs.available<2Mi,imagefs.available<2Mi" \
+                --extra-config=kubelet.eviction-soft-grace-period="memory.available=2h,nodefs.available=2h,imagefs.available=2h"
```
how do `image-gc-X-threshold` / `ttl-duration` related to disk space problems?

rmetzger replied:
Not at all. This relates to the second problem I described above `Problem: Low disk space let the kublet delete unused docker images, including images generated for the test.`.

zentol replied:
You say they aren't related to low disk space, but the problem you refer to is read by me as `Low disk space [caused issues].`

I'm gonna need a longer explanation :/

rmetzger replied:
The VMs provided by azure have 100GB of disk space, out of which 85% are allocated, only are 15GB are free.
That's enough space for our purposes.
However, the kubernetes nodes running during the k8s tests believe that 85% are not enough free disk space, so they start garbage collecting their host.
During GCing, they are deleting all docker images currently not in use.
However, the k8s test is first building a flink image, then launching stuff on k8s. Sometimes, k8s deletes the newly Flink images, thus it can not find them anymore, letting the test fail / timeout.

That's why I have set the GC threshold to 98% and 99% :)

rmetzger replied:
Other people have a similar collection of configuration parameters https://github.com/swaroopar/automation/blob/c1643b1cb66e6591ec33c9e8569ad5721b824052/minikube/start-minikube.sh :)

zentol replied:
Thanks for the explanation.

Now document it :P

rmetzger replied:
Done

zentol commented on tools/azure-pipelines/jobs-template.yml:
```diff
@@ -177,6 +145,8 @@ jobs:
     - script: ./tools/azure-pipelines/setup_kubernetes.sh
     - script: M2_HOME=/home/vsts/maven_cache/apache-maven-3.2.5/ PATH=/home/vsts/maven_cache/apache-maven-3.2.5/bin:$PATH ${{parameters.environment}} STAGE=compile ./tools/azure_controller.sh compile
       displayName: Build
+    - script: FLINK_DIR=build-target ./flink-end-to-end-tests/run-pre-commit-tests.sh
```
add a TODO to remove this since we're pretty much eliminating the concept of pre-commit e2e tests.

rmetzger replied:
Very good point, thanks!

zentol commented on tools/azure-pipelines/build-apache-repo.yml:
```diff
@@ -48,3 +58,60 @@ stages:
+          container: flink-build-jdk8
+      - template: jobs-template.yml
+        parameters:
+          stage_name: cron_build_scala2_12
```
would skip to naming convention of the hadoop profile, i.e. `cron_build_scala212`

zentol commented on tools/azure-pipelines/build-apache-repo.yml:
```diff
@@ -48,3 +58,60 @@ stages:
+          environment: PROFILE=""
+          run_end_to_end: true
+          container: flink-build-jdk8
+      - job: docs_404_check # run on a MSFT provided machine
```
where are you caching the gem artifacts?

rmetzger replied:
I'm not caching them at all. But this job passes in 5 minutes, so I'm not concerned.

zentol replied:
it's not a matter of build times but of stability and resource conservation.

zentol commented on tools/azure-pipelines/jobs-template.yml:
```diff
@@ -132,20 +134,26 @@ jobs:
         path: $(CACHE_FLINK_DIR)
         artifact: FlinkCompileCacheDir-${{parameters.stage_name}}
+    - script: |
+        echo "##vso[task.setvariable variable=JAVA_HOME]$(JAVA_HOME_11_X64)"
```
Is this to have the tests run on java 11 if java 11 is available on the machine? Shouldn't the container have this stuff set up already (or in other words, what's the benefit in having distinct images if extra steps are required to select the jdk version)?

rmetzger replied:
This step is only used on the azure machines (where we do not use the docker image).
That's why I'm changing the java version like this.

zentol replied:
But ... there is condition for this job(?) that ${{parameters.container}} must be `flink-build-jdk11`. Are we _not_ using `rmetzger/flink-ci:ubuntu-jdk11-amd64-2a765ab` when running on azure machines?

rmetzger replied:
Correct, this is a bit ugly. I basically decide whether we are using jdk11 or not based on the container name (that is not used in this context).

I can address this as follows: I build a generic docker image (that does not distinguish between jdk8 / jdk11 because it contains both), and I switch the jdk version for docker and azure machines the same way.

Are you okay if I address this in a follow up? (I can only test this during the night, because I would overwhelm the available machines)

zentol commented on tools/azure_controller.sh:
```diff
@@ -172,6 +172,7 @@ elif [ $STAGE != "$STAGE_CLEANUP" ]; then
         PY_MVN="${MVN// clean/}"
         PY_MVN="$PY_MVN -Drat.skip=true"
         ${PY_MVN}
+        EXIT_CODE=$?
```
unrelated

zentol commented on tools/azure-pipelines/jobs-template.yml:
```diff
@@ -159,9 +167,15 @@ jobs:
-    - script: M2_HOME=/home/vsts/maven_cache/apache-maven-3.2.5/ PATH=/home/vsts/maven_cache/apache-maven-3.2.5/bin:$PATH PROFILE="-Dinclude-hadoop -Dhadoop.version=2.8.3 -De2e-metrics -Dmaven.wagon.http.pool=false" STAGE=compile ./tools/azure_controller.sh compile
+    - script: M2_HOME=/home/vsts/maven_cache/apache-maven-3.2.5/ PATH=/home/vsts/maven_cache/apache-maven-3.2.5/bin:$PATH ${{parameters.environment}} STAGE=compile ./tools/azure_controller.sh compile
       displayName: Build
     - script: FLINK_DIR=`pwd`/build-target flink-end-to-end-tests/run-nightly-tests.sh
```
where are you running the java e2e tests?

rmetzger replied:
In my understanding these are executed in the `misc` profile, from the `travis_watchdog.sh` script. (See an example of a `misc` profile: https://dev.azure.com/rmetzger/Flink/_build/results?buildId=5590&view=logs&j=b09cc737-3452-5710-2b65-ee4e507c2164&t=8141c33c-fb34-5bae-c47a-dcb67e7c446b)

zentol replied:
only the pre-commit e2e tests, which is 1 out of 4 right now.

rmetzger commented:
Thank you for your review so far!
I have addressed all comments, except for the 3/4 java end to end tests and the caching of the docs 404 check.
Would you be okay with merging this PR as is, and I address these two issues in another PR?
I am not sure if I will manage to implement, test and review those items today, but I would like to have the nightlies and e2e test execution in master asap, because keeping the tests somewhat running / stable is a moving target

zentol commented:
After an offline chat with @rmetzger we concluded that the caching for the doc gems may not be necessary.

I'm fine with delaying the inclusion of java e2e tests; at this point you want to load test things anyway and they don't _that_ much additional overhead. Additionally we would need a convenient hook that enables all e2e tests.

The jdk11 container stuff is indeed icky, but I'm fine with merging it as is _for the time being_; it's certainly not state we can stay in for a longer period.

rmetzger commented:
Thanks a lot. I will merge this change then!
